plugins/outputs/postgresql/postgresql.go

```diff
@@ -0,0 +1,424 @@
+	if err != nil {
+		return fmt.Errorf("starting transaction: %w", err)
+	}
+	defer tx.Rollback(p.dbContext)
+
+	for _, tableSource := range tableSources {
+		err := p.writeMetricsFromMeasure(p.dbContext, tx, tableSource)
+		if err != nil {
+			if isTempError(err) {
+				return err
+			}
+			p.Logger.Errorf("write error (permanent, dropping sub-batch): %v", err)
+		}
+	}
+
+	if err := tx.Commit(p.dbContext); err != nil {
+		return fmt.Errorf("committing transaction: %w", err)
+	}
+	return nil
+}
+
+func (p *Postgresql) writeConcurrent(tableSources map[string]*TableSource) error {
+	for _, tableSource := range tableSources {
+		select {
+		case p.writeChan <- tableSource:
+		case <-p.dbContext.Done():
+			return nil
+		}
+	}
+	return nil
+}
+
+func (p *Postgresql) writeWorker(ctx context.Context) {
+	defer p.writeWaitGroup.Done()
+	for {
+		select {
+		case tableSource, ok := <-p.writeChan:
+			if !ok {
+				return
+			}
+			if err := p.writeRetry(ctx, tableSource); err != nil {
+				p.Logger.Errorf("write error (permanent, dropping sub-batch): %v", err)
+			}
+		case <-p.dbContext.Done():
+			return
+		}
+	}
+}
+
+// This is a subset of net.Error
+type maybeTempError interface {
+	error
+	Temporary() bool
+}
+
+// isTempError reports whether the error received during a metric write operation is temporary or permanent.
+// A temporary error is one that if the write were retried at a later time, that it might succeed.
+// Note however that this applies to the transaction as a whole, not the individual operation. Meaning for example a
+// write might come in that needs a new table created, but another worker already created the table in between when we
+// checked for it, and tried to create it. In this case, the operation error is permanent, as we can try `CREATE TABLE`
+// again and it will still fail. But if we retry the transaction from scratch, when we perform the table check we'll see
+// it exists, so we consider the error temporary.
+func isTempError(err error) bool {
+	var pgErr *pgconn.PgError
+	if errors.As(err, &pgErr); pgErr != nil {
+		// https://www.postgresql.org/docs/12/errcodes-appendix.html
+		errClass := pgErr.Code[:2]
+		switch errClass {
+		case "42": // Syntax Error or Access Rule Violation
+			switch pgErr.Code {
+			case "42701": // duplicate_column
+				return true
+			case "42P07": // duplicate_table
+				return true
+			}
+		case "53": // Insufficient Resources
+			return true
+		case "57": // Operator Intervention
+			return true
+		case "23": // Integrity Constraint Violation
+			switch pgErr.Code {
+			case "23505": // unique_violation
+				if strings.Contains(err.Error(), "pg_type_typname_nsp_index") {
+					// Happens when you try to create 2 tables simultaneously.
+					return true
+				}
+			}
+		}
+		// Assume that any other error that comes from postgres is a permanent error
+		return false
+	}
+
+	if mtErr := maybeTempError(nil); errors.As(err, &mtErr) {
+		return mtErr.Temporary()
+	}
```

Comment:
Do you know of errors beside `net` errors that have that function? If not, you might directly use the net.Error interface here.